srtcore/threadname.h

```diff
@@ -40,22 +36,61 @@ class ThreadName
         return prctl(PR_SET_NAME, (unsigned long)name, 0, 0) != -1;
     }
 
+    ThreadName(const char* name)
+    {
+        if ((good = get(old_name)))
+            set(name);
+    }
+
+    ~ThreadName()
+    {
+        if (good)
+            set(old_name);
+    }
+
+private:
+    char old_name[BUFSIZE];
+    bool good;
+};
+
+#elif defined(__APPLE__)
+#include <pthread.h>
+#include <cstring>
+
+class ThreadName
+{
+public:
+    static const size_t BUFSIZE = 64;
+
+    static bool get(char* namebuf)
+    {
+        return pthread_getname_np(pthread_self(), namebuf, BUFSIZE) == 0;
+    }
+
+    static bool set(const char* name)
+    {
+        char tmp[BUFSIZE];
+        tmp[BUFSIZE - 1] = '\0';
+        std::strncpy(tmp, name, BUFSIZE - 1);
+        return pthread_setname_np(tmp) == 0;
+    }
```

Comment:
Some thoughts on potential improvement. Although it is a minor thing.

Maybe using just `strlen` would be simpler? It expects null termination, but all those names are internal to the library and do have the termination character.
As an option the `set` function could take either `const std::string&` or an additional length parameter.

Reply:
This is only to prevent a string from being infinite because when `strncpy` reaches the given size, it doesn't insert the NUL character at the end.

Reply:
> Some thoughts on potential improvement. Although it is a minor thing.
> 
> Maybe using just `strlen` would be simpler? It expects null termination, but all those names are internal to the library and do have the termination character.
> As an option the `set` function could take either `const std::string&` or an additional length parameter.

My original purpose is to mimic the behavior of prctl: if the length of the string, including the terminating null byte, exceeds N bytes, the string is silently truncated. Otherwise pthread_setname_np() would fail if name is too long, although it's unlikely.

I'm fine to remove the check if you prefer.